### PR TITLE
Restore log step during restart

### LINKED ISSUE #12274: Resuming training resets the logged step number
## 🐛 Bug

The change introduced in https://github.com/PyTorchLightning/pytorch-lightning/pull/11805 causes a reset to the logged step number.
https://github.com/PyTorchLightning/pytorch-lightning/blob/49a4a36ad45b937dd0124ecfb08eb7400dbf3950/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py#L122

### To Reproduce
```
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks import ModelCheckpoint


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run(ckpt_path=None):
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        num_sanity_val_steps=0,
        max_epochs=2,
        enable_model_summary=False,
        callbacks=ModelCheckpoint(dirpath="checkpoints", save_top_k=-1, filename="{epoch}", save_on_train_epoch_end=False),
        log_every_n_steps=1,
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data, ckpt_path=ckpt_path)


if __name__ == "__main__":
    run()
    run("checkpoints/epoch=0.ckpt")
```
The script will create two tensorboard logs:
- version_0: steps 0 to 63
- version_1: steps 0 to 31

### Expected behavior
- version_1: steps 31 to 63

This was the behavior before https://github.com/PyTorchLightning/pytorch-lightning/pull/11805

### Environment
- PyTorch Lightning Version (e.g., 1.5.0): master ([49a4a36)](https://github.com/PyTorchLightning/pytorch-lightning/commit/49a4a36ad45b937dd0124ecfb08eb7400dbf3950)
- Fault-tolerant training is off (PL_FAULT_TOLERANT_TRAINING=0)

cc @tchaton @rohitgr7 @akihironitta @awaelchli @ananthsub @ninginthecloud @carmocca

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #12274

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy @justusschock @ninginthecloud

### CODE REVIEW

File: src/pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 1]
```diff
@@ -273,6 +273,7 @@ def teardown(self) -> None:
 
     def on_save_checkpoint(self) -> Dict:
         state_dict = super().on_save_checkpoint()
+        state_dict["batches_that_stepped"] = self._batches_that_stepped
```
Comment (carmocca) [inline src/pytorch_lightning/loops/epoch/training_epoch_loop.py:276]: I would keep the underscore

```suggestion
        state_dict["_batches_that_stepped"] = self._batches_that_stepped
```

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 2]
```diff
@@ -292,6 +293,7 @@ def on_save_checkpoint(self) -> Dict:
     def on_load_checkpoint(self, state_dict: Dict) -> None:
         # cache the dataloader state dict until the dataloader objects are available
         self._dataloader_state_dict = state_dict.get("dataloader_state_dict")
+        self._batches_that_stepped = state_dict.get("batches_that_stepped")
```
Comment (rohitgr7) [inline src/pytorch_lightning/loops/epoch/training_epoch_loop.py:296]:
```suggestion
        self._batches_that_stepped = state_dict.get("_batches_that_stepped")
```

File: tests/tests_pytorch/loops/test_loop_state_dict.py [hunk 3]
```diff
@@ -47,7 +47,7 @@ def test_loops_state_dict_structure():
     expected = {
         "fit_loop": {
             "state_dict": {},
-            "epoch_loop.state_dict": {},
+            "epoch_loop.state_dict": {"batches_that_stepped": 0},
```
Comment (rohitgr7) [inline tests/tests_pytorch/loops/test_loop_state_dict.py:50]:
```suggestion
            "epoch_loop.state_dict": {"_batches_that_stepped": 0},
```

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 4]
```diff
@@ -292,6 +293,7 @@ def on_save_checkpoint(self) -> Dict:
     def on_load_checkpoint(self, state_dict: Dict) -> None:
         # cache the dataloader state dict until the dataloader objects are available
         self._dataloader_state_dict = state_dict.get("dataloader_state_dict")
+        self._batches_that_stepped = state_dict.get("_batches_that_stepped")
```
Comment (awaelchli) [inline src/pytorch_lightning/loops/epoch/training_epoch_loop.py:296]: How is mypy passing this test? _batches_that_stepped here is an int, but get() can return None.

```suggestion
        self._batches_that_stepped = state_dict.get("_batches_that_stepped", 0)
```

**Review by awaelchli — CHANGES_REQUESTED** [review]

**Review by awaelchli — APPROVED** [review]
